CromwellRefdiskManifestCreator/src/main/java/org/broadinstitute/manifestcreator/CromwellRefdiskManifestCreatorApp.java

```diff
@@ -0,0 +1,151 @@
+package org.broadinstitute.manifestcreator;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.core.config.Configurator;
+import org.broadinstitute.manifestcreator.exception.CRC32Exception;
+import org.broadinstitute.manifestcreator.model.ReferenceDiskManifest;
+import org.broadinstitute.manifestcreator.model.ReferenceFile;
+
+import java.io.*;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Stack;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.stream.Stream;
+import java.util.zip.CRC32;
+
+public class CromwellRefdiskManifestCreatorApp {
+
+  private static final Logger logger = LogManager.getLogger(CromwellRefdiskManifestCreatorApp.class);
+
+  public static void main(String[] args) throws IOException, InterruptedException {
+    Configurator.setRootLevel(Level.DEBUG);
+
+    if (args.length < 4) {
+      logger.error("Wrong number of parameters.");
+      logger.error("Usage: java -jar cromwell-refdisk-manifest-creator-app.jar <number of parallel threads> " +
+              "<image_name> <directory_path_to_scan> <output file path");
+      System.exit(1);
+    }
+
+    String directoryToScan = args[2];
+    File rootDirectory = new File(directoryToScan);
+    if (!rootDirectory.exists() || !rootDirectory.isDirectory()) {
+      logger.error("Root directory " + directoryToScan + " doesn't exist.");
+      System.exit(1);
+    }
+
+    int totalNumberOfFiles = countFiles(rootDirectory);
+    logger.info("Parsing " + totalNumberOfFiles + " files");
+
+    ReferenceDiskManifest manifest = new ReferenceDiskManifest();
+    String imageName = args[1];
+    manifest.setImageIdentifier(imageName);
+
+    int nThreads = Integer.parseInt(args[0]);
+    CountDownLatch countDownLatch = new CountDownLatch(nThreads);
+    ExecutorService executorService = Executors.newFixedThreadPool(nThreads);
+
+    Path rootPath = rootDirectory.toPath();
+    Stack<File> stack = new Stack<>();
+    stack.push(rootDirectory);
+    List<ReferenceFile> referenceFiles = Collections.synchronizedList(new ArrayList<>());
```

Comment:
Is the `Stack` assumed to be synchronized by default? Is that why we don't need to wrap it?

Reply:
Yes, stack is a thread-safe wrapper (adding `push` and `pop` methods) around `Vector` (which is also thread-safe).